
# Exploring Hackers News Posts
In this project, we'll compare two different types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:
* Do Ask HN or Show HN receive more comments on average? 
* Do posts created at a certain time receive more comments on average?

It should be noted that the [data set](https://www.kaggle.com/hacker-news/hacker-news-posts) we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Introduction
First, we'll read in the data and remove the headers.

In [1]:
from csv import reader
open_file=open('hacker_news.csv')
read_file=reader(open_file)
hn=list(read_file)
header=hn[0]
hn=hn[1:]
print(header)
print(hn[:6])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'], ['10482257', 'Ti

First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [2]:
ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title=row[1]
    if (title.lower()).startswith('ask hn'):
        ask_posts.append(row)
    elif (title.lower()).startswith('show hn'):
        show_posts.append(row)
    else: 
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [3]:
print(header)
print(ask_posts[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [4]:
total_ask_comments=0
for row in ask_posts:
    num_comments=int(row[4])
    total_ask_comments+=num_comments
avg_ask_comments=total_ask_comments/len(ask_posts)
print('Avg number of comments on ask posts:',avg_ask_comments)

total_show_comments=0
for row in show_posts:
    num_comments=int(row[4])
    total_show_comments+=num_comments
avg_show_comments=total_show_comments/len(show_posts)
print('Avg number of comments on ask posts:',avg_show_comments)

Avg number of comments on ask posts: 14.038417431192661
Avg number of comments on ask posts: 10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

# Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [5]:
import datetime as dt

result_list=[]
for row in ask_posts:
    list=[]
    created_at=row[6]
    list.append(created_at)
    num_com=int(row[4])
    list.append(num_com)
    result_list.append(list)
    
counts_by_hour={}
comments_by_hour={}
dt_format="%m/%d/%Y %H:%M"

for row in result_list:
    date=row[0]
    comments=row[1]
    time=dt.datetime.strptime(date,dt_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time]=1
        comments_by_hour[time]=comments
    else:
        counts_by_hour[time]+=1
        comments_by_hour[time]+=comments
        
comments_by_hour

    

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [6]:
avg_by_hour=[]
for hour in counts_by_hour:
    l=[]
    l.append(hour)
    l.append(comments_by_hour[hour]/counts_by_hour[hour])
    avg_by_hour.append(l)
avg_by_hour    

[['20', 21.525],
 ['15', 38.5948275862069],
 ['03', 7.796296296296297],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['23', 7.985294117647059],
 ['05', 10.08695652173913],
 ['18', 13.20183486238532],
 ['00', 8.127272727272727],
 ['17', 11.46],
 ['01', 11.383333333333333],
 ['02', 23.810344827586206],
 ['09', 5.5777777777777775],
 ['11', 11.051724137931034],
 ['16', 16.796296296296298],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['19', 10.8],
 ['21', 16.009174311926607],
 ['08', 10.25],
 ['14', 13.233644859813085],
 ['07', 7.852941176470588],
 ['22', 6.746478873239437],
 ['12', 9.41095890410959]]

In [7]:
def takeSecond(elem):
    return elem[1]

k=1
sort_list=sorted(avg_by_hour, key=takeSecond, reverse=True)
sort_list


[['15', 38.5948275862069],
 ['02', 23.810344827586206],
 ['20', 21.525],
 ['16', 16.796296296296298],
 ['21', 16.009174311926607],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['18', 13.20183486238532],
 ['17', 11.46],
 ['01', 11.383333333333333],
 ['11', 11.051724137931034],
 ['19', 10.8],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['12', 9.41095890410959],
 ['06', 9.022727272727273],
 ['00', 8.127272727272727],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['03', 7.796296296296297],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['09', 5.5777777777777775]]

# Sorting and Printing Values from a List of Lists

In [10]:
print("Top 5 Hours for 'Ask HN' Comments")
for row in sort_list[:5]:
    h=row[0]
    c=row[1]
    hr=dt.datetime.strptime(h,"%H")
    hour=hr.strftime("%H:%M")
    template="{}: {:.2f} average comments per post"
    output=template.format(hour,c)
    print(output)
    
    

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.